00_Base/src/interfaces/messages/MessageRouter.ts

```diff
@@ -14,12 +13,113 @@ import { IClientConnection } from "../centralsystem";
+
+    private _cache: ICache;
+    private _sender: IMessageSender;
+    private _handler: IMessageHandler;
+
+    constructor(cache: ICache, sender: IMessageSender, handler: IMessageHandler) {
+        this._cache = cache;
+        this._sender = sender;
+        this._handler = handler;
+    }
+
+    async routeCall(connectionIdentifier: string, message: Call): Promise<IMessageConfirmation> {
+        const messageId = message[1];
+        const action = message[2] as CallAction;
+        const payload = message[3] as OcppRequest;
+
+        // TODO: Add tenantId to context
+        const context: IMessageContext = { correlationId: messageId, stationId: connectionIdentifier, tenantId: '' };
+
+        // TODO: Use base util builder instead
+        const _message: IMessage<OcppRequest> = {
+            origin: MessageOrigin.ChargingStation,
+            eventGroup: EventGroup.General, // TODO: Change to appropriate event group
+            action,
+            state: MessageState.Request,
+            context,
+            payload
+        };
+
+        return this._sender.send(_message);
+    }
+
+    async routeCallResult(connectionIdentifier: string, message: CallResult, action: CallAction): Promise<IMessageConfirmation> {
+        const messageId = message[1];
+        const payload = message[2] as OcppResponse;
+
+        // TODO: Add tenantId to context
+        const context: IMessageContext = { correlationId: messageId, stationId: connectionIdentifier, tenantId: '' };
+
+        const _message: IMessage<OcppRequest> = {
+            origin: MessageOrigin.CentralSystem,
+            eventGroup: EventGroup.General,
+            action,
+            state: MessageState.Response,
+            context,
+            payload
+        };
+
+        return this._sender.send(_message);
+    }
+
+    async routeCallError(connectionIdentifier: string, message: CallError, action: CallAction): Promise<IMessageConfirmation> {
+        const messageId = message[1];
+        const payload = new OcppError(messageId, message[2], message[3], message[4]);
+
+        // TODO: Add tenantId to context
+        const context: IMessageContext = { correlationId: messageId, stationId: connectionIdentifier, tenantId: '' };
+
+        const _message: IMessage<OcppError> = {
+            origin: MessageOrigin.CentralSystem,
+            eventGroup: EventGroup.General,
+            action,
+            state: MessageState.Response,
+            context,
+            payload
+        };
+
+        // Fulfill callback for api, if needed
+        this.handleMessageApiCallback(_message);
+
+        // No error routing currently done
+        throw new Error('Method not implemented.');
+    }
+
+    async handleMessageApiCallback(message: IMessage<OcppError>): Promise<void> {
+        const url: string | null = await this._cache.get(message.context.correlationId, this.CALLBACK_URL_CACHE_PREFIX + message.context.stationId);
```

Comment:
There's a separate cache entry that contains the correlationId of Calls that haven't received CallResults yet (see CacheNamespace.Transactions use in CentralSystem class). This entry is used to send responses or associated messages to whoever fired the original trigger. Right now that is just CallResults and CallErrors, but I'm hoping the behavior can be extended for Reports and Transactions (so that Notify messages and TransactionEvents/MeterValues can be sent to a callback if they were triggered originally though the message API). To support that, I don't think we'll be able to remove the callback from the cache proactively because there will be an unknown number of total messages that could be sent to it.